lib/Sema/CSSimplify.cpp

```diff
@@ -4302,33 +4302,51 @@ getDynamicCallableMethods(Type type, ConstraintSystem &CS,
   auto it = CS.DynamicCallableCache.find(canType);
   if (it != CS.DynamicCallableCache.end()) return it->second;
 
-  auto calculate = [&]() -> DynamicCallableMethods {
-    // If this is a protocol composition, check if any of the protocols have the
-    // attribute.
-    if (auto protocolComp = dyn_cast<ProtocolCompositionType>(canType)) {
-      DynamicCallableMethods methods;
-      for (auto protocolType : protocolComp->getMembers()) {
-        auto tmp = getDynamicCallableMethods(protocolType, CS, locator, error);
-        if (error) return methods;
-        if (tmp.argumentsMethod) {
-          if (methods.argumentsMethod &&
-              methods.argumentsMethod != tmp.argumentsMethod) {
-            error = true;
-            return methods;
-          }
-          methods.argumentsMethod = tmp.argumentsMethod;
+  // Calculate @dynamicCallable methods for composite types with multiple
+  // components (protocol composition types and archetypes).
+  auto calculateForComponentTypes =
+      [&](ArrayRef<Type> componentTypes) -> DynamicCallableMethods {
+    DynamicCallableMethods methods;
+    for (auto componentType : componentTypes) {
+      auto tmp = getDynamicCallableMethods(componentType, CS, locator, error);
+      if (error) return methods;
+      if (tmp.argumentsMethod) {
+        if (methods.argumentsMethod &&
+            methods.argumentsMethod != tmp.argumentsMethod) {
+          error = true;
+          return methods;
         }
-        if (tmp.keywordArgumentsMethod) {
-          if (methods.keywordArgumentsMethod &&
-              methods.keywordArgumentsMethod != tmp.keywordArgumentsMethod) {
-            error = true;
-            return methods;
-          }
-          methods.keywordArgumentsMethod = tmp.keywordArgumentsMethod;
+        methods.argumentsMethod = tmp.argumentsMethod;
+      }
+      if (tmp.keywordArgumentsMethod) {
+        if (methods.keywordArgumentsMethod &&
+            methods.keywordArgumentsMethod != tmp.keywordArgumentsMethod) {
+          error = true;
+          return methods;
         }
+        methods.keywordArgumentsMethod = tmp.keywordArgumentsMethod;
       }
-      return methods;
     }
+    return methods;
+  };
+
+  // Calculate @dynamicCallable methods.
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have the attribute.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
```

Comment:
@slavapestov Could you verify whether the code here for handling archetypes is sufficient?

To get `@dynamicCallable` methods from an archetype, I look for the methods in the archetype's `getConformsTo()` protocols and superclass (if defined). Are there cases that aren't handled by this?

Reply:
This looks reasonable.

At this point, the 'component types' calculation and superclass walk duplicates work that's being done in name lookup. It would be great if at some later point in time we could come up with a better abstraction here!

Reply:
Thanks for the astute observation!
The "type walking" logic in `getDynamicCallableMethods` is gnarly and should definitely be fixed + simplified if possible. I'm not sure how to best do this at the moment.